xapian-letor/include/xapian-letor/ranker.h

```diff
@@ -354,6 +354,68 @@ class XAPIAN_VISIBILITY_DEFAULT SVMRanker: public Ranker {
 
 };
 
+class XAPIAN_VISIBILITY_DEFAULT ListMLERanker: public Ranker {
+    /// Ranker parameters
+    std::vector<double> parameters;
+    /// Learning rate (Default is 0.001)
+    double learning_rate;
+    /// Number of iterations (Default is 10)
+    int iterations;
+
+    /** Train the model.
+     *
+     * @exception InvalidArgumentError will be thrown if training_data is
+     *            empty.
+     */
+    void train(const std::vector<Xapian::FeatureVector> & training_data);
+
+    /** Save ListMLE model as db metadata.
+     *
+     *  ListMLE model file gets stored with each parameter value in a new line.
+     *	e.g.
+     *
+     *  0.000920817564536697
+     *  0.000920817564536697
+     *  0
+     *  -1.66533453693773e-19
+     *
+     *  @param model_key	Metadata key using which model is to be stored.
+     */
+    void save_model_to_metadata(const std::string & model_key);
+
+    /** Load model from an external file.
+     *
+     *  @param model_key         Metadata key using which model is to be
+     *				 loaded.
+     *
+     *  @exception LetorInternalError will be thrown if no model exists
+     *  	   corresponding to the supplied key
```

Comment:
`InvalidArgumentError` seems more appropriate here (and below).

The documentation of `LetorInternalError` is rather vague as to its intended purpose ("LetorInternalError indicates a runtime problem of some sort"), but given the name I'd expect it to be for things like unexpected errors from libsvm.